src/treelearner/serial_tree_learner.cpp

```diff
@@ -172,18 +173,29 @@ Tree* SerialTreeLearner::Train(const score_t* gradients, const score_t *hessians
   int cur_depth = 1;
   // only root leaf can be splitted on first time
   int right_leaf = -1;
-  for (int split = 0; split < tree_config_->num_leaves - 1; ++split) {
+
+  int init_splits = 0;
+  bool aborted = false;
+  if (!forced_split_json.is_null()) {
+    init_splits = ForceSplits(tree.get(), forced_split_json, &left_leaf,
+                              &right_leaf, &cur_depth, &aborted);
+  }
+
+  for (int split = init_splits; split < tree_config_->num_leaves - 1; ++split) {
     #ifdef TIMETAG
     start_time = std::chrono::steady_clock::now();
     #endif
     // some initial works before finding best split
-    if (BeforeFindBestSplit(tree.get(), left_leaf, right_leaf)) {
+    if (!aborted && BeforeFindBestSplit(tree.get(), left_leaf, right_leaf)) {
```

Comment:
what is `aborted` used for ?

Reply:
Aborted is set if force splits aborts - if the gain from the forced split is worse than without the split. At this point, BeforeFindBestSplit and FindBestSplits have already been called in ForceSplits, so they can't be called again in the first iteration of the main training loop. This is because the functions aren't idempotent - calling them twice leads to wrong behavior. Passing in an aborted flag was the best way to detect this without modifying the existing code.

Reply:
got it, thanks very much

Reply:
maybe rename `aborted` to `aborted_last_force_split` ?